server/src/main/java/org/elasticsearch/common/logging/HeaderWarningLogger.java

```diff
@@ -0,0 +1,337 @@
+        for (int i = 0; i < s.length(); i++) {
+            final char c = s.charAt(i);
+            if (c == '\\' || c == '"') {
+                escapingNeeded = true;
+                break;
+            }
+        }
+
+        if (escapingNeeded) {
+            final StringBuilder sb = new StringBuilder();
+            for (final char c : s.toCharArray()) {
+                if (c == '\\' || c == '"') {
+                    sb.append("\\");
+                }
+                sb.append(c);
+            }
+            return sb.toString();
+        } else {
+            return s;
+        }
+    }
+
+    /**
+     * Encode a string containing characters outside of the legal characters for an RFC 7230 quoted-string.
+     *
+     * @param s the string to encode
+     * @return the encoded string
+     */
+    static String encode(final String s) {
+        // first check if the string needs any encoding; this is the fast path and we want to avoid creating a string builder and copying
+        boolean encodingNeeded = false;
+        for (int i = 0; i < s.length(); i++) {
+            int current = s.charAt(i);
+            if (doesNotNeedEncoding.get(current) == false) {
+                encodingNeeded = true;
+                break;
+            }
+        }
+
+        if (encodingNeeded == false) {
+            return s;
+        }
+
+        final StringBuilder sb = new StringBuilder(s.length());
+        for (int i = 0; i < s.length(); ) {
+            int current = s.charAt(i);
+            /*
+             * Either the character does not need encoding or it does; when the character does not need encoding we append the character to
+             * a buffer and move to the next character and when the character does need encoding, we peel off as many characters as possible
+             * which we encode using UTF-8 until we encounter another character that does not need encoding.
+             */
+            if (doesNotNeedEncoding.get(current)) {
+                // append directly and move to the next character
+                sb.append((char) current);
+                i++;
+            } else {
+                int startIndex = i;
+                do {
+                    i++;
+                } while (i < s.length() && doesNotNeedEncoding.get(s.charAt(i)) == false);
+
+                final byte[] bytes = s.substring(startIndex, i).getBytes(UTF_8);
+                // noinspection ForLoopReplaceableByForEach
+                for (int j = 0; j < bytes.length; j++) {
+                    sb.append('%').append(hex(bytes[j] >> 4)).append(hex(bytes[j]));
+                }
+            }
+        }
+        return sb.toString();
+    }
+
+    private static char hex(int b) {
+        final char ch = Character.forDigit(b & 0xF, 16);
+        if (Character.isLetter(ch)) {
+            return Character.toUpperCase(ch);
+        } else {
+            return ch;
+        }
+    }
+
+    public void log(String msg, Object... params) {
```

Comment:
can this be `warn` instead of log ?

Reply:
hm not sure about this one. Wouldn't some verb be better here? addHeader? addHeaderWarning? 
or maybe simply log as it is now. The class name and the field name should express the intent